### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by Anntonia**

### CODE REVIEW

File: CountingIslands.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import static org.junit.Assert.assertEquals;
+import org.junit.Test;
+
```
Comment (nfortescue) [inline CountingIslands.java]: Little nit comment: This would be much easier to read in the github web interface if you changed your tab indent  to less than 8 spaces. Maybe 2?

File: CountingIslands.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+public class CountingIslands {
+	
+	public int countIslands(int rows, int columns, boolean[][] map){
+		int numberOfIslands = 0;
```
Comment (nfortescue) [inline CountingIslands.java]: Every level of indent you add to a method makes if a little bit harder to read. Not just physically, but every indent is a mental note I have to make, either an extra loop variable, or an condition, etc. 

I'd give yourself a mental rule than whenever you indent more than 3 or 4 times, think about extracting a helper method.  It will make your code much more readable.

File: CountingIslands.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	    boolean[][] map = new boolean[4][4];
+	    map[0][1] = true;
+	    map[0][3] = true;
+	    map[1][0] = true;
+	    map[1][1] = true;
+	    map[2][2] = true;
+	    map[3][2] = true;
+	    assertEquals(countIslands(4,4,map), 3);
+	  }
+
```
Comment (nfortescue) [inline CountingIslands.java]: You need a much more complete set of test cases here. This is not enough to convince me your algorithm works.
Maybe have separate test methods:
- a single island in a corner
- a single island in the opposite corner (to check edge conditions)
- an island where they are adjacent horizontally (to check that merge)
- an island where they are adjacent vertically (to check that merges)
- a chessboard style grid where they are all disjoint, but adjacent diagonally.

By checking each of these individually, if the test fails you know where a bug was introduced. You also convince the reader who is thinking "but what about ..."

File: CountingIslands.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		for(int i = 0; i < rows; i++){
+			for(int j = 0; j < columns; j++){
+				if(map[i][j]){
+					if(i > 0){
+						if(j > 0){
+							if(!map[i-1][j] && !map[i][j-1]){
+								numberOfIslands++;
+							}
+							else if(map[i-1][j] && map[i][j-1]){
+								numberOfIslands--;
```
Comment (nfortescue) [inline CountingIslands.java]: I had to really, really stare at your code to work out how the two if statements were different. Maybe you could have added some helper methods to make it more readable. For example, you could have introduced the idea of northCoast (change from sea in the row above land), southCoast, eastCoast and westCoast.
Then your code would be:
if (land and not north or west edge of board) {
  if northCoast & westCoast: number++
  else if inlandNorth & inlandWest: number--
  else if northCoast: number++
  else if westCoast: number++
}

This doesn't look like it can be right. Suppose we had a grid:

{false, false}
{false, true}
{false, true}
{false, true}
{false, true}

This is 1 island, but your code looks like it counts it as 5?

Also, it should be a red flag there is a bug that you always ignore land in column 0. But 
{true, false} should give 1 island, and {false, false} should give 0.

File: CountingIslands.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import static org.junit.Assert.assertEquals;
+import org.junit.Test;
+
+public class CountingIslands {
+	
+	public int countIslands(int rows, int columns, boolean[][] map){
+		int numberOfIslands = 0;
+		for(int i = 0; i < rows; i++){
```
Comment (nfortescue) [inline CountingIslands.java]: If you are iterating over rows or columns, instead of calling the loop variables i & j, you could call them r & c, or row & column. Then the code is easier to read.
